esp-hal/src/dma/mod.rs

```diff
@@ -1851,6 +1850,23 @@ where
     ) -> Result<(), DmaError> {
         debug_assert_eq!(preparation.direction, TransferDirection::In);
 
+        if !unsafe { &*preparation.start }.iter().all(|d| {
+            if preparation.burst_transfer.is_burst_enabled() && is_valid_ram_address(d.buffer as _)
+            {
+                return preparation.burst_transfer.can_burst_in(d);
+            }
+
+            #[cfg(psram_dma)]
+            if is_valid_psram_address(d.buffer as _) {
+                return self.rx_impl.can_access_psram()
+                    && preparation.burst_transfer.external.can_burst_in(d);
+            }
+
+            true
+        }) {
+            return Err(DmaError::InvalidAlignment);
+        }
+
```

Comment:
This is a very aggressive check. The Dma Buffer traits being unsafe meant these could be skipped.

In my project using "what has now become the DPI driver", I could have up to 100,000 descriptors for a single frame buffer. This would kill my application's rendering performance.

Maybe only do this for the `DescriptorChain` code path?

Reply:
> I could have up to 100,000 descriptors for a single frame buffer.

That is what, 1.2MB in SRAM just for descriptors? What device are you using?

Reply:
Whoops, apologies, wrong number. That's the size of the diff buffer. It's 5,000 descriptors.

Reply:
Alright I can try and come up with something. Agressive or not, alignment and location requirements depend on what ends up transmitting a buffer, but maybe we can optimize for more special cases.

Reply:
Ideally there should just be some way for buffers to express what they requirements they conform to in the `Preparation` struct.

The preliminary P4 datasheet barely goes into detail, so idk what additional requirements are needed there. esp-idf is littered with L1 cache invalidation calls before DMA operations, so there's more fun there I guess.

Reply:
My main worry is that the user sets up a buffer with word alignment, then tries to send it via the AXI DMA which has perma-enabled burst mode with a minimum of 8-byte (configurable) block size. We can't reject this in the buffer constructor, because we don't know that the AXI DMA will be used, and the AHB DMA doesn't need so strict alignment. Then, the user can end up with weird chunk sizes as well, because 4092 isn't divisible by 8.

Even worse, the DMA you need depends on the peripheral, not the memory location - the AXI DMA is used by SPI for example, regardless of internal or external RAM.

I think the best we can do is to overalign our data to the strictest requirements present, and just deal with that. The AXI DMA even requires 2-word descriptor alignment which is extra funny considering they are 3 words each.

Reply:
> AXI DMA which has perma-enabled burst mode with a minimum of 8-byte (configurable) block size

Noted. Are you allowed to share the possible configurations? I guess I could check the PAC/esp-idf but I'd rather not guess haha.

I guess `is_burstable` being a `bool` doesn't work anymore, as it's not a chip-wide setting on the P4.
It now needs to be some kind of `struct`/`enum` describing the byte alignment. Tough.

In the name of optional complexity, maybe `is_burstable` as bool can remain on non-P4 chips, and P4 chips would have the `struct`/`enum` stuff. Any buffer libraries would then have conditional compilation.

> I think the best we can do is to overalign our data to the strictest requirements present, and just deal with that. The AXI DMA even requires 2-word descriptor alignment which is extra funny considering they are 3 words each.

Yeah the buffer implementations in the hal should probably do this, however users writing their own bufs may not want to deal with such strict requirements on their non-P4 chip.

Given the P4 is practically a wrecking ball to the `Preparation` stuff, perhaps it should be `#[non_exhaustive]`.
Then users just do `Preparation::with_descriptor(...)` to start building one.
I didn't think it'd be this different :sweat_smile:.

> I think the best we can do is to overalign our data to the strictest requirements present, and just deal with that. The AXI DMA even requires 2-word descriptor alignment which is extra funny considering they are 3 words each.

Lmao, wtf. What a waste of space....

Can't wait to see what you cook up :smile:

Reply:
> perhaps it should be #[non_exhaustive]

Do you expect it to change over time, in a given chip?

Reply:
> > perhaps it should be #[non_exhaustive]
> 
> Do you expect it to change over time, in a given chip?

Ah good point. Not really. Depends on the design decisions made here I guess.
If `Preparation` is made chip-agnostic then it's going to change with each new innovative chip.

Reply:
> If Preparation is made chip-agnostic then it's going to change with each new innovative chip.

I don't think that's a problem in terms of exhaustiveness - I don't think libraries should assume APIs remain stable even between different chips. I believe if they want to extend esp-hal, they will need to provide the same chip features we do. We can try our best, but some things may not end up being hardware agnostic. @MabezDev what do you think?

Reply:
fwiw I do prefer if it's not hardware agnostic. Makes everyone's lives easier imo.